Feature/preserve comment only files

marcules opened the pull request:
<!-- Describe what the change is, trying to link it with open issues -->
Fixes #183 

Adds a temporary key to the document at the start and removes it again after ruyaml ran through, preventing it from removing the comments in a comment-only yaml file.

## Checklist

* [x] Add test cases to all the changes you introduce
* ~[ ] Update the documentation for the changes~

marcules commented:
@lyz-code @rsnodgrass this adds a mapping key to the document before it goes through ruyaml - meaning:
1. if the document already has mapping keys on the top level, this line is treated as if it were a mapping with a "none" value
2. if the document already has a sequences as the top level, this line is treated as if it were the name of the sequence

When it has other mappings `yamlfix_document_fix_id` has none value:
```yaml
---
yamlfix_document_fix_id: # uuid
key: value
---
# equivalent to:
yamlfix_document_fix_id: none # uuid
key: value
```

When it has a top level list, `yamlfix_document_fix_id` is the name of said list:
```yaml
---
yamlfix_document_fix_id: #uuid
- item
- item
---
# equivalent to:
yamlfix_document_fix_id: [item, item] # uuid
```

When the document was previously only comments, ruyaml attaches them to the "last found node" (e.g. this line):
```yaml
---
yamlfix_document_fix_id: #uuid
# other comments
# comments comments
```

All previous tests are passing, and this should in theory work with any yaml file - but it is, admittedly, "clever" (not to say hacky 😅) - not exactly sure what happens with shorter files, that are converted to flow-style or something like that.

marcules commented:
@lyz-code please wait on merging this until #199 is merged and this is rebased. I want to modify it a bit.

rsnodgrass commented:
Thanks @marcules for the fix!!

lyz-code commented:
@marcules #199 is merged!

lyz-code commented:
@marcules pull the changes from `main` and run `make install`. I've updated the dependencies versions

marcules commented:
@lyz-code you can review this and pull it in now it it looks good
I have rebased it onto main and also fixed the issue in regards to flakeheaven looking into folders other than src/ tests/

lyz-code commented on docs/index.md:
```diff
@@ -200,6 +200,18 @@ export YAMLFIX_CONFIG_PATH="/etc/yamlfix/"
+Warning: You should not modify this value, if you're not sure you need to.
+
+This generates a UUID in hex-string-representation (no delimiters), which is used internally to generate a temporary top-level mapping node, where any lists or comments can be attached, so ruyaml is not removing comments, and comment-only documents can be formatted properly.
+
```
Do you think this is something a user would want to change? I don't feel that without knowing the internals of `yamlfix` they can even understand the paragraph.

If you feel it's interesting to keep the configuration, I'd move it to the bottom of the document, otherwise I'd just hardcode the value

lyz-code commented on src/yamlfix/adapters.py:
```diff
@@ -252,6 +259,9 @@ def patch_sequence_style(key_node: Node, value_node: Node) -> None:
         self.patch_functions.append(patch_sequence_style)
 
+    def _seq_is_in_top_level_node(self, key_node: Node) -> bool:
+        return str(key_node.value) == f"yamlfix_{self.config.document_fix_id}"
```
Following the question above, can't we just use `yamlfix_top_node`?

lyz-code commented on src/yamlfix/adapters.py:
```diff
@@ -693,3 +705,33 @@ def _restore_jinja_variables(source_code: str) -> str:
+    def _fix_comment_only_files(self, source_code: str) -> str:
+        """Add a mapping key with an id to the start of the document\
+            to preserve comments."""
+        fixed_source_lines = []
```
Now we're running `_fix_comment_only_files` and `_restore_comment_only_files` by default, which slows down the program as it needs to iterate through the source code 2 times. 

I'd only run these functions if the config is set.

lyz-code commented on src/yamlfix/adapters.py:
```diff
@@ -693,3 +705,33 @@ def _restore_jinja_variables(source_code: str) -> str:
+            fixed_source_lines.append(line)
+            if line.startswith("---"):
+                has_start_indicator = True
+                fixed_source_lines.append(yamlfix_document_id_line)
```
For performance reasons it would be nice if we could stop the iteration on the file source as soon as we've done the required changes. Same applies at the `_restore_comment_only_files`.

Maybe reformulate it so that once the condition is met it injects the required line and exits. 

Now that I see it, although this is a good idea, it may need a non-trivial refactor where the fixers tweak directly the parts of the code from `self.` instead of receiving the source as an argument and returning as a return value.

If you don't feel like doing this now we can open an issue to track it for the future

lyz-code reviewed (changes requested):
Thanks again for your time, do you want to become a maintainer of the package? I can give you permissions, that way we can share the load of answering the users, taking care of the CI...

If you don't like the idea it's fine :)
